### DISCUSSION TITLE
Translations Manual (translators wanted!)

### DISCUSSION THREAD

**Discussion by AlexHTW** [n3d1117/chatgpt-telegram-bot#219] [Q&A]
# Currently available languages
You can set your preferred `BOT_LANGUAGE` code in the .env file.

| Language  | Code | Translator(s) |
| ------------- | ------------- | -------------|
|  🇬🇧 English 🇬🇧| `en`  | _default_ |
| 🇩🇪 German 🇩🇪 | `de`  | @AlexHTW |
| 🇷🇺 Russian 🇷🇺 | `ru`  | @AlexHTW |
| 🇹🇷 Turkish 🇹🇷 | `tr`  | @muhammed540 |
| 🇮🇹 Italian 🇮🇹 | `it`  | @n3d1117 |
| 🇮🇩 Indonesian 🇮🇩 | `id`  | @Am1nCmd |
| 🇪🇸 Spanish 🇪🇸 | `es`  | @jvican |
| 🇳🇱 Dutch 🇳🇱 | `nl`  | @BjornB2 |
| 🇨🇳 Simplified Chinese 🇨🇳 | `zh-cn`  | @deanxizian |
| 🇹🇼 Traditional Chinese 🇹🇼 | `zh-tw`  | @aes-alienrip |
| 🇻🇳 Vietnamese 🇻🇳 | `vi`  | @HoangAnhQuoc |
| 🇮🇷 Farsi (Persian) 🇮🇷 | `fa`  | @bugfloyd |
| 🇫🇮 Finnish 🇫🇮 | `fi`  | @jnaskali |
| 🇧🇷 Brazilian Portuguese 🇧🇷  | `pt-br`  | @gkpiccoli |
| 🇺🇦 Ukrainian 🇺🇦  | `uk`  | @nmeln |
| 🇲🇾 Malay 🇲🇾  | `ms`  | @Yaekirua |
| 🇺🇿 Uzbek 🇺🇿  | `uz`  | @mirmakhamat |
| 🇸🇦 Arabic 🇸🇦 | `ar`  | @EyadMahm0ud |
| 🇵🇱 Polish 🇵🇱 | `pl`  | @Rafael-6fx |
| 🇮🇱 Hebrew 🇮🇱 | `he`  | @ShadovvBeast |

# Contribute (no coding necessary)

## Add a new language
- submit a PR with your translations added to the `translations.json` and update the README.md file OR
- make a comment on this thread translating the current list of texts (see below) into your language OR
- submit your translations in [this form](https://docs.google.com/forms/d/e/1FAIpQLSd6PGEZCyicuOpUamAT6_ubEFuaw2jyMR8PLgx-iyQk_9MVDA/viewform?usp=sf_link) and someone else can put them into a PR

Mind variables before or after a text snippet in your translation so that it will read correctly in the complete text.
For example the text `This month your OpenAI account was billed $` should be translated in a way that the actual number can come after the text. If a literal translation doesn't work in your language, feel free to adjust as necessary (for example change it to something like `Your OpenAI bill this month: $`).

## Reviewing translations
Submit a PR, comment on this thread or comment on a cell in the [google sheet ](https://docs.google.com/spreadsheets/d/1ovkrVjzXUMHcW2yWSt-tAvg_YqnQAX8xydQH4RfKWys/edit?usp=sharing) to suggest improvements to existing translations.

# Current list of texts 
Check the [translations.json file](https://github.com/n3d1117/chatgpt-telegram-bot/blob/main/translations.json) to see all implemented translations.
See the [google sheet](https://docs.google.com/spreadsheets/d/1ovkrVjzXUMHcW2yWSt-tAvg_YqnQAX8xydQH4RfKWys/edit?usp=sharing) for a formatted view. It is updated hourly from the main branch.

# Adding new bot messages (for code contributors)
When submitting new features that include new bot messages in the telegram chat, add the text content (without variables and formatting) to the English part of the `translations.json` with a descriptive and unique key. In the code retrieve the text using the `localized_text('YOUR_NEW_KEY', self.config['bot_language'])` function. 
If your text is split up by formatting or variables you can use a list for that key, see for example "help_text" or "stats_transcribe".

You can fill in translations for all languages using tools such as deepl, chatgpt or copilot. If you don't add the translations it will default to the English texts which is also okay.
It would probably speed up things to tag the existing translators in the PR or comment about your update on this thread for translators to review or add the new bot messages.

# Questions and Suggestions
Let me know if anything is unclear or you have improvement suggestions. I am looking forward to your translations 🌍.

**Comment by khanfar** [comment]
hi , can you add Arabic Language ?

**Reply by AlexHTW** [reply]
Hey, I don't speak Arabic but I'd be happy to add it to the code if you fill in the translations in a new column of the Google sheet 😇

**Reply by khanfar** [reply]
oh ok

**Reply by mnchrmxd** [reply]
Done

**Comment by n3d1117** [comment]
Hi @AlexHTW, just merged your PR and pinned this discussion! 🚀 

While I was adding italian translation I noticed a small issue with the `stats_budget` translation:
```
"stats_budget":"Your remaining budget for",
"monthly":"this month",
"daily":"today",
"all-time":"",
```

When budget was set to `all-time`, the text would look like: `Your remaining budget for : $X.X`

So I changed all translations to:
```
"stats_budget":"Your remaining budget",
"monthly":" for this month",
"daily":" for today",
"all-time":"",
```
and removed the hardcoded space from `text_budget`, so that it becomes `Your remaining budget: $X.X`

Hope you're ok with it! (changes are in #222)

PS: Not a big fan of the google doc (too much duplication imo), but if you'd like to maintain it that's great! (if so, please update it with the changes above)

**Reply by AlexHTW** [reply]
Hey @n3d1117, you're right, good catch and thanks for fixing it! 
I will adjust the translations accordingly.

@muhammed540
Could you let me know how the Turkish translation should be changed when we move the "for" from "Your remaining budget" to " for this month" and " for today"? Thanks!

> PS: Not a big fan of the google doc (too much duplication imo), but if you'd like to maintain it that's great! (if so, please update it with the changes above)

I understand your concern, it will be challenging to always keep it updated and of course I can not guarantee how long I will be interested in doing that😅) .
I will add another option in the main text to just provide the translations here in this thread and keep the google doc as a bonus and to keep track for now. My original intention was to provide an easier interface for new translations. The ultimate source of truth is of course the `translations.json`.
_(hmmm 💡 maybe I can figure out how to pull the data from there into the doc so the part for existing translations will always be correct and automatically updated...)_

**Reply by AlexHTW** [reply]
@n3d1117 You already fixed German and Russian! Nice :)

**Reply by AlexHTW** [reply]
> (hmmm bulb maybe I can figure out how to pull the data from there into the doc so the part for existing translations will always be correct and automatically updated...)

Soooo, this sent me on quite a tangent this weekend. I discovered Apps Scripts which you can use to automate all kinds of google docs stuff. I implemented a script to sync the [sheet](https://docs.google.com/spreadsheets/d/1ovkrVjzXUMHcW2yWSt-tAvg_YqnQAX8xydQH4RfKWys/edit?usp=sharing) with the translations.json and the script is run hourly. You can see the code [here](https://github.com/AlexHTW/google-apps-scripts/blob/main/translations_json_import.gs). I tested it for all kinds of changes and so far no issues, as long as the link to the translations.json stays the same.

Additionally I added a script that is turning the elements of the "en" column into a [form](https://docs.google.com/forms/d/e/1FAIpQLSd6PGEZCyicuOpUamAT6_ubEFuaw2jyMR8PLgx-iyQk_9MVDA/viewform?usp=sf_link), so that the form can be used for people to submit new translations. I set this script to run once a day, because it takes longer and I want to minimize the chance of an update during a submission. Here is the [source code](https://github.com/AlexHTW/google-apps-scripts/blob/main/create_form_from_sheet_column.gs) for that. The submissions can be found in the third tab of the sheet.

I updated the main text above.
This was fun and I am really happy nothing needs any manual input now, except to check for new submissions _(hmmm I could also connect the submissions table to github api with Apps Scripts and automatically comment the new translation here....😅 )_

**Reply by n3d1117** [reply]
Awesome @AlexHTW! 😃

**Comment by n3d1117** [comment]
New strings added:
```
"answer_with_chatgpt":"Answer with ChatGPT",
"ask_chatgpt":"Ask ChatGPT",
"loading":"Loading..."
```

`de`, `ru` and `tr` are translated with Copilot, please open a PR if they look wrong!

**Comment by samdoghor** [comment]
Can I add Nigeria Pidgin English?

**Reply by AlexHTW** [reply]
Why not?  If it's useful to you, it might be useful to others as well 🙂

**Reply by samdoghor** [reply]
Thank you

**Comment by aes-alienrip** [comment]
Hi, @n3d1117 and @AlexHTW 
Traditional Chinese mainly used in Taiwan, Hong Kong and Macau but not Mainland China.
IMO flags are actually symbols of countries, not of languages.

**Reply by n3d1117** [reply]
Hey there, thanks for clarifying. What would be the correct flags to use here?

**Reply by AlexHTW** [reply]
I'm fine with removing them everywhere. I just thought it looked nicer 😅 but I understand that it's not very inclusive for many languages.

**Reply by aes-alienrip** [reply]
Hi, @n3d1117 
For `zh-tw` :taiwan: instead of :cn: if persist in using flags to represent languages

**Reply by n3d1117** [reply]
Updated!

**Comment by erhanurgun** [comment]
Add TR language support for README.md

Hello,

Your bot looks very nice and very useful. Although Python is not my area of interest, I felt the need to do something again for such a nice repo.

While thinking about what I could do, I came up with the idea of translating README.md file according to the language rules for the Turkish language.

If you allow me, I would like to create a folder named lang for this repository, create a file named readme-tr.md, and translate your README.md file into Turkish.

**Comment by KevinBandara** [comment]
Add cooldown system to the Bot. like with 3 secs of cool down time per message, and it will decrease the Bot resources usage and also tokens

**Comment by ilya-88** [comment]
Ребят я только начинаю этим заниматься поэтому не судите строго в параметрах Бота я изменил язык бота на русский но приветствие всё равно на английском выводится что я делаю не так # BOT_LANGUAGE=ru

**Comment by KevinBandara** [comment]
oh are u russian?

**Comment by ilya-88** [comment]
Ес оф корс , я уже разобрался как ,не  совсем правильно но работает

**Comment by KevinBandara** [comment]
I love russia from sri lanka. btw are u using discord??
